### PR TITLE
Use container of nc datetime and calendar.

### PR DESCRIPTION
**Pull request by lbdreyer**
To integrate this with Iris, we want to make use of the newer version of netcdf4.

New netcdf4 has slightly different behaviour, e.g. differing number results; not allowing you to add a calendar attribute to netcdftime.datetime object (See [PR6 comment ](https://github.com/SciTools/nc-time-axis/pull/6#issuecomment-229050703) for more detail)

This adds a container for the netcdftime.datetime object and associated calendar object, then registers that container with matplotlib.units instead.

### CODE REVIEW

File: nc_time_axis/__init__.py [hunk 1]
```diff
@@ -233,14 +248,20 @@ def convert(cls, value, unit, axis):
                 return value
             first_value = value
 
-        if not hasattr(first_value, 'calendar'):
-            raise ValueError('A "calendar" attribute must be attached to '
-                             'netcdftime object to understand them properly.')
-        return netcdftime.date2num(value, cls.standard_unit,
-                                   calendar=first_value.calendar)
+        if not isinstance(first_value.datetime, netcdftime.datetime):
```
Comment (bjlittle) [inline nc_time_axis/__init__.py]: @lbdreyer I guess at this point you're expected the `first_value` to be a `CalendarDateTime`, otherwise the `first_value.datetime` might/will raise an `AttributeError` exception. Did you want to also consider that here? Is that appropriate?

File: nc_time_axis/__init__.py [hunk 2]
```diff
@@ -233,14 +248,20 @@ def convert(cls, value, unit, axis):
...
+            raise ValueError('The datetime attribute of the CalendarDateTime '
+                             'object must be of type `netcdftime.datetime`.')
+
+        ut = netcdftime.utime(cls.standard_unit, calendar=first_value.calendar)
+
+        if isinstance(value, Iterable):
+            num = ut.date2num([v.datetime for v in value])
+        else:
+            num = ut.date2num(value.datetime)
```
Comment (bjlittle) [inline nc_time_axis/__init__.py]: @lbdreyer If you are expecting `value` to be an `Iterable` of `CalendarDateTime` or a `CalendarDateTIme` instance, you could alternatively do the following ...
```python
if isinstance(value, CalendarDateTime):
    value = [value]

return ut.date2num([v.datetime for v in value])
```

File: nc_time_axis/__init__.py (hunk 1 above)
Comment (lbdreyer) [inline nc_time_axis/__init__.py]: I think I was counting (incorrectly) on the fact that someone would never use `NetCDFTimeConvert` directly, that it was only there for matplotlib to make use of. Since `NetCDFTimeConvert` had been registered with in the `munits.regristry` with the object `CalendarDateTime`, I thought I could assume that a `CalendarDateTime` object would only ever be passed to this.
I think I'm wrong in that assumption so I'm going to add a check that it is of type `CalendarDateTime`.

File: nc_time_axis/__init__.py (hunk 2 above)
Comment (bjlittle) [inline nc_time_axis/__init__.py]: @lbdreyer It might be that this is an artificial case (for a testing perspective) and when plugged into matplotlib we'd never see this. Perhaps you could confirm and we'll take it from there.

Comment (lbdreyer) [inline nc_time_axis/__init__.py]: So... looking at matplotlib ...
In matplotlib/axis.py, matplotlib makes use of this `convert` method to convert the tick values and to set the default intervals. In the setting of the default intervals, it is passing a single element to the `convert` method and uses the result to set the `axes.dataLim.intervalx` and `axes.viewLim.intervalx` (and the same for y).

I would think it would be a problem in places like [here](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_axes.py#L2129)... but it doesn't seem to be  

When testing the values that come out, for normal numbers you get:
```
>>> plt.plot([1, 2, 3])
[<matplotlib.lines.Line2D object at 0x24d9510>]
>>> plt.gca().viewLim.intervaly
array([ 1.,  3.])
```

for netcdftime.datetimes:
```
>>> dates = [nc_time_axis.CalendarDateTime(netcdftime.datetime(1986, month, 30), '360_day') for month in range(1, 6)]
>>> line1, = plt.plot(dates)
>>> plt.gca().viewLim.intervaly
array([ -5.01100000e+03,   3.00000000e+00])
```

Basically it doesn't seem to be a problem

File: nc_time_axis/__init__.py [hunk 3]
```diff
@@ -248,17 +248,21 @@ def convert(cls, value, unit, axis):
                 return value
             first_value = value
 
+        if not isinstance(first_value, CalendarDateTime):
+            msg = 'The value must be a number, a sequence of numbers or of ' \
+                  'type :class:`nc_time_axis.CalendarDateTime.'
```
Comment (bjlittle) [inline nc_time_axis/__init__.py]: @lbdreyer Dah! You're missing the closing tick around `nc_time_axis.CalendarDateTime`.

You could avoid the whole issue and instead say:
`... sequence of numbers or an instance of "nc_time_axis.CalendarDateTime"`

The usage of `:class:...` is Sphinx-like markup that won't be rendered differently, but shown raw as-is when this exception is raised.

Change either way, your choice :wink:

Comment (lbdreyer) [inline nc_time_axis/__init__.py]: In fact this is only checking the type of the first value which should be a number or `CalendarDateTime` and not a sequence at all

**Comment by bjlittle** [comment]
@lbdreyer Looking good :+1:
